src/Rings/MPolyMap/flattenings.jl

```diff
@@ -0,0 +1,247 @@
+########################################################################
+# Computation of kernels of homomorphisms of polynomial rings 
+# over coefficient rings which are not fields.
+#
+# In the admissible cases, this flattens the structure of the rings 
+# and computes the kernel on the flat counterpart. 
+########################################################################
+function kernel(
+    f::MPolyAnyMap{<:AbstractAlgebra.Generic.MPolyRing{<:RingElemType}, 
+                   <:AbstractAlgebra.Generic.MPolyRing{<:RingElemType}
+                  }
+  ) where {RingElemType <: Union{<:MPolyElem, <:MPolyQuoElem, <:MPolyLocalizedRingElem, 
+                                 <:MPolyQuoLocalizedRingElem
+                                }
+          }
+  f_flat = _flatten(f)
+  K_flat = kernel(f_flat)
+  K = ideal(domain(f), _flat_to_orig(domain(f)).(gens(K_flat)))
+  #set_attribute!(K, :_flat_counterpart, K_flat) # TODO: Support attribute storage?
+  return K
+end
+
+@attr function _flatten(
```

Comment:
Can you construct the return type from the type of `S`? If so, we could add it here for type stability.

Reply:
Yes, that would be good. It's a bit lengthy, but possible.